### ISSUE TITLE
Enhancement Request: Better Cargo Packing

### ISSUE THREAD

**Issue by marphod** [endless-sky/endless-sky#2754]
I'm not sure what the cargo packing algorithm is, but it does not seem to be one that optimizes for the storage of large items.  At a guess, it places items in the first available cargo hold with enough space.

A better packing algorithm would be:
For items to store (largest to smallest) 
If there is a cargo hold with exactly item.size space left, use that
otherwise, place in hold with currently the most available space

This will avoid cases where my 160 ton cargo ships had a 90 ton object and random trade goods, rather than the 150 ton objects i wanted them to have.

**Comment by tehhowch** [comment]
<details><summary>Current packing is done as mission cargo -> spare outfits -> commodities
</summary>


PlayerInfo::TakeOff (PlayerInfo.cpp#L1168)
```
for(const shared_ptr<Ship> &ship : ships)
		if(!ship->IsParked() && ship->GetSystem() == system && !ship->IsDisabled())
		{
			ship->Recharge(hasSpaceport);
			if(ship != flagship)
			{
				ship->Cargo().SetBunks(ship->Attributes().Get("bunks") - ship->RequiredCrew());
				cargo.TransferAll(&ship->Cargo());
			}
			else
			{
				// Your flagship takes first priority for passengers but last for cargo.
				ship->Cargo().SetBunks(ship->Attributes().Get("bunks") - ship->Crew());
				for(const auto &it : cargo.PassengerList())
					cargo.TransferPassengers(it.first, it.second, &ship->Cargo());
			}
		}
	// Load up your flagship last, so that it will have space free for any
	// plunder that you happen to acquire.
	cargo.TransferAll(&flagship->Cargo());
```
The relevant bits TransferAll routine (CargoHold.cpp#L457)
```
	for(const auto &it : passengers)
		TransferPassengers(it.first, it.second, to);
	// Handle zero-size mission cargo correctly. For mission cargo, having an
	// entry in the map, but with a size of zero, is different than not having
	// an entry at all.
	auto mit = missionCargo.begin();
	while(mit != missionCargo.end())
	{
		Transfer(mit->first, mit->second, to);
		if(!mit->second)
			mit = missionCargo.erase(mit);
		else
			++mit;
	}
	for(const auto &it : outfits)
		Transfer(it.first, it.second, to);
	for(const auto &it : commodities)
		Transfer(it.first, it.second, to);
```
and lastly, Transfer (outfits version):
```
{
	double mass = outfit->Get("mass");
	
	// Whichever ship is removing the cargo is limited by how much it has
	// available. The receiving ship is limited by its free space, but only if
	// its space is limited (i.e. size is not set to -1).
	amount = min(amount, Get(outfit));
	if(size >= 0 && mass)
		amount = max<int>(amount, -max(Free(), 0) / mass);
	if(to)
	{
		amount = max(amount, -to->Get(outfit));
		if(to->size >= 0 && mass)
			amount = min<int>(amount, max(to->Free(), 0) / mass);
	}
	if(!amount)
		return 0;
	
	// The "to" hold need not be defined.
	outfits[outfit] -= amount;
	if(to)
		to->outfits[outfit] += amount;
	
	return amount;
}
```
</details>

I think perhaps all that needs to be done is to change the sorting operator for the `outfits` data container, such that it sorts from largest to smallest mass (or use a different container that will auto-sort such that it can be appropriately iterated). Then when it is filling, it will attempt to fill ships in decreasing size order, which means your largest outfits go into the ship first ship with space for them, and then only smaller outfits are checked.

**Comment by tehhowch** [comment]
If/when #2757 merges, we can enjoy better output packing in about 90% of launches than the current method of outfit packing. See my comment/gist: https://github.com/endless-sky/endless-sky/pull/2757#issuecomment-315519576

**Comment by tehhowch** [comment]
This can be closed with the `0.9.7 (done)` milestone
